win32ss/gdi/ntgdi/dibobj.c

```diff
@@ -1242,147 +1250,255 @@ NtGdiStretchDIBitsInternal(
+            return 0;
+        }
 
-    if (pdc->fs & (DC_ACCUM_APP|DC_ACCUM_WMGR))
-    {
-       IntUpdateBoundsRect(pdc, &rcDst);
-    }
+        /* Select the bitmap into hdcMem, and save a handle to the old bitmap */
+        hOldBitmap = NtGdiSelectBitmap(hdcMem, hBitmap);
 
-    hbmTmp = GreCreateBitmapEx(pbmi->bmiHeader.biWidth,
-                               abs(pbmi->bmiHeader.biHeight),
-                               0,
-                               BitmapFormat(pbmi->bmiHeader.biBitCount,
-                                            pbmi->bmiHeader.biCompression),
-                               pbmi->bmiHeader.biHeight < 0 ? BMF_TOPDOWN : 0,
-                               cjMaxBits,
-                               pvBits,
-                               0);
+        if (dwUsage == DIB_PAL_COLORS)
+        {
+            hPal = NtGdiGetDCObject(hdc, GDI_OBJECT_TYPE_PALETTE);
+            hPal = GdiSelectPalette(hdcMem, hPal, FALSE);
+        }
+
+        if (pbmi->bmiHeader.biCompression == BI_RLE4 ||
+                pbmi->bmiHeader.biCompression == BI_RLE8)
+        {
+            /* copy existing bitmap from destination dc */
+            if (cxSrc == cyDst && cySrc == cyDst)
+            {
+                NtGdiBitBlt(hdcMem, xSrc, abs(pbmi->bmiHeader.biHeight) - cySrc - ySrc,
+                            cxSrc, cySrc, hdc, xDst, yDst,  dwRop, 0, 0);
+            }
+            else
+            {
+                NtGdiStretchBlt(hdcMem, xSrc, abs(pbmi->bmiHeader.biHeight) -     cySrc - ySrc,
+                                cxSrc, cySrc, hdc, xDst, yDst, cxDst, cyDst,
+                                dwRop, 0);
+            }
+        }
+
+        pdc = DC_LockDc(hdcMem);
+        if (pdc != NULL)
+        {
+            IntSetDIBits(pdc, hBitmap, 0, abs(pbmi->bmiHeader.biHeight), safeBits,
+                         cjMaxBits, pbmi, dwUsage);
+            DC_UnlockDc(pdc);
+        }
+
+        /* Origin for DIBitmap may be bottom left (positive biHeight) or top
+           left (negative biHeight) */
+        if (cxSrc == cxDst && cySrc == cyDst)
+        {
+            NtGdiBitBlt(hdc, xDst, yDst, cxDst, cyDst,
+                        hdcMem, xSrc, abs(pbmi->bmiHeader.biHeight) - cySrc - ySrc,
+                        dwRop, 0, 0);
+        }
+        else
+        {
+            NtGdiStretchBlt(hdc, xDst, yDst, cxDst, cyDst,
+                            hdcMem, xSrc, abs(pbmi->bmiHeader.biHeight) - cySrc - ySrc,
+                            cxSrc, cySrc, dwRop, 0);
+        }
+
+        /* cleanup */
+        if (hPal)
+            GdiSelectPalette(hdcMem, hPal, FALSE);
+
+        if (hOldBitmap)
+            NtGdiSelectBitmap(hdcMem, hOldBitmap);
+
+        NtGdiDeleteObjectApp(hdcMem);
+
+        GreDeleteObject(hBitmap);
+
+    cleanup1:
+
+        ExFreePoolWithTag(safeBits, TAG_DIB);
+
+    } /* End of dwRop == SRCCOPY */
+    else
+    { /* Start of dwRop != SRCCOPY */
```

Comment:
Seeing these two large cases I am wondering whether having two helper functions, one for dwRop == SRCCOPY, and another one for dwRop != SRCCOPY, would be better.... ? 🤔

Reply:
Hopefully for another round of updates with someone who have more energy than I do. :)

Reply:
Thanks for the suggestion, @HBelusca. Are you volunteering? :)

Reply:
Then later on...